py/obj.h

```diff
@@ -590,8 +603,162 @@ struct _mp_obj_type_t {
+#define _MP_DEFINE_CONST_OBJ_TYPE_12(_typename, _name, _flags, _make_new, f1, v1, f2, v2, f3, v3, f4, v4, f5, v5, f6, v6, f7, v7, f8, v8, f9, v9, f10, v10, f11, v11, f12, v12) const mp_obj_type_t _typename = { .base = { &mp_type_type }, .name = _name, .flags = _flags | (_MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f1 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f2 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f3 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f4 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f5 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f6 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f7 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f8 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f9 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f10 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f11 | _MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f12) * MP_OBJ_TYPE_FLAG_EXTENDED, .make_new = _make_new, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f1 = v1, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f2 = v2, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f3 = v3, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f4 = v4, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f5 = v5, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f6 = v6, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f7 = v7, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f8 = v8, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f9 = v9, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f10 = v10, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f11 = v11, ._MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f12 = v12 }
+
+#define MP_OBJ_TYPE_HAS_SLOT(t, f) (_MP_DEFINE_CONST_OBJ_TYPE_SLOT_IS_EXT_##f ? (((t)->flags & MP_OBJ_TYPE_FLAG_EXTENDED) ? (t)->_MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f : 0) : (t)->_MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f)
+#define MP_OBJ_TYPE_GET_SLOT(t, f) ((t)->_MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f)
+#define MP_OBJ_TYPE_GET_SLOT_OR_NULL(t, f) ((MP_OBJ_TYPE_HAS_SLOT(t, f) ? MP_OBJ_TYPE_GET_SLOT(t, f) : NULL))
+#define MP_OBJ_TYPE_SET_SLOT(t, f, v, n) ((t)->_MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f = v)
+#define MP_OBJ_TYPE_OFFSETOF_SLOT(t, f) (offsetof(mp_obj_type_t, _MP_DEFINE_CONST_OBJ_TYPE_SLOT_PATH_##f))
+
+#endif
+
+// Workaround for https://docs.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview?view=msvc-160#macro-arguments-are-unpacked
+#define _MP_DEFINE_CONST_OBJ_TYPE_EXPAND(x) x
+
+#define _MP_DEFINE_CONST_OBJ_TYPE_IMPL(_1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, N, ...) _MP_DEFINE_CONST_OBJ_TYPE_##N
+#define MP_DEFINE_CONST_OBJ_TYPE(...) _MP_DEFINE_CONST_OBJ_TYPE_EXPAND(_MP_DEFINE_CONST_OBJ_TYPE_IMPL(__VA_ARGS__, _INV, 12, _INV, 11, _INV, 10, _INV, 9, _INV, 8, _INV, 7, _INV, 6, _INV, 5, _INV, 4, _INV, 3, _INV, 2, _INV, 1, _INV, 0, _INV, _INV, _INV)(__VA_ARGS__))
```

Comment:
I was curious how you would handle this. I looked into how boost preprocessor (the one I was familiar with) accomplishes the things it does, like dispatching based on the number of arguments, and _nope_'d right back out of there as quickly as I could.  This isn't too bad, though.

Reply:
> I see you rediscovered the problem with directly initializing slots inside ext[0].

Yeah it's fortunate that GCC warns but does the right thing anyway. I also can't figure out exactly what the pattern to how it decides which fields to warn against is.

> This isn't too bad, though.

Thanks. I was very relieved when this worked across gcc/clang/msvc, and TBH I don't think (other than msvc's quirk with expanding __VA_ARGS__) there's anything too controversial here. It also works on older gcc (6.4) and icc and tcc. It would be bad news if this ruled out another compiler though... iar is the main one I haven't tried.